### ISSUE TITLE
[FEATURE] Dynamic number and geometry of widgets?

### ISSUE THREAD

**Issue by ckp95** [elkowar/eww#187]
## Description of the requested feature
I'm not sure if this is already possible, if it is please can you tell me how to do it?

I'd like to use eww to dynamically put labels or other thingies on top of open windows. This is so I can create e.g. a window switcher. I would press a hotkey, which would trigger a script to get the geometries of the open windows, calculate their centre points, then tell the eww daemon to bring up a bunch of boxes with letters inside at those coordinates. Then it would listen for keypresses of those keys, switch focus to the appropriate window, and tell the eww daemon to remove the boxes.

This can already be done in a few window managers. For example, there is ["revelation"](https://github.com/guotsuan/awesome-revelation) in AwesomeWM. Screenshot of what I'm talking about [here](https://raw.githubusercontent.com/guotsuan/awesome-revelation/master/screenshot.png). But I want to be able to implement this in a WM-agnostic way.

To be clear, I'm not asking how to make eww find window geometries or switch focus or anything, I can already do that with any number of other tools. I just want to be able to create little boxes where the number and placement and contents are only known at runtime, not at "compile time" (config time?)

## Proposed configuration syntax
I don't really know :x

**Comment by undefinedDarkness** [comment]
Yes you can as for geometry other than anchor, width height will be calculated by eww dynamically iirc

**Comment by ckp95** [comment]
> Yes you can as for geometry other than anchor, width height will be calculated by eww dynamically iirc

Can you explain how? I can't figure out how to use `literal` to create multiple windows.

**Comment by elkowar** [comment]
Currently, given that literals are only widgets, not windows, they cannot be used to generate windows arbitrarily. While you can influence the position and geometry of windows dynamically, its currently not possible to generate a dynamic amount of windows.

This usecase is definitely something I'd want to support in the future! 
What I will probably do to fulfill this is add a `eww create-window` command like this:
```
eww create-window \
  --name="foo" \
  --geometry="200,400,50,50" \
  --widget="<box>foo</box>"
```
Which would allow creating windows completely dynamically. 

I fear however that this will take a bit, as there are multiple larger changes in progress right now that have higher priority in my mind.

**Comment by ckp95** [comment]
>While you can influence the position and geometry of windows dynamically, its currently not possible to generate a dynamic amount of windows.

I see. In that case I suppose I could workaround it by just hardcoding like a hundred windows (more than anyone would conceivably have open in normal operation) and then just dynamically alter the visibility and positions of the widgets they contain via script.

**Comment by elkowar** [comment]
Yea, until I implement this feature, that is sadly gonna be the only real workaround

**Comment by ckp95** [comment]
>While you can influence the position and geometry of windows dynamically

How do I do this? I can't get it to use the variable I defined.

I have this:

```xml
<eww>
    <definitions>    
        <def name="simple_box">
            <box>i'm a box!</box>
        </def>
    </definitions>

    <variables>
        <var name="x_pos">300px</var>
    </variables>

    <windows>    
        <window name="simple_box" screen="0" >
            <geometry x="500px" y="900px" width="200px" height="200px"/>
            <widget >
                <simple_box/>
            </widget>
        </window>
    </windows>
</eww>
```

This displays a box at the specified coordinates as it should. But if I do this:

```xml
<geometry x="{{x_pos}}" y="900px" width="200px" height="200px"/>
```

It doesn't work. When I use `eww logs` I get:

```
(eww:2128): Gtk-WARNING **: 15:25:35.294: Theme parsing error: <data>:9:15: Not using units is deprecated. Assuming 'px'.
 2021-06-05T14:25:35.294Z ERROR eww::server     > Failed to reload config: Failed to load eww config file /home/cameron/.config/eww/eww.xml

Caused by:
    0: Error parsing eww config file /home/cameron/.config/eww/eww.xml
    1: Error parsing window definition at /home/cameron/.config/eww/eww.xml:13:9
    2: Parsing the value of x="{{x_pos}}" in <geometry>
    3: could not parse '{{x_pos}}'
```

I've tried other ways of quoting it but none seem to work. How do I get the variable into the `geometry` tag?

**Comment by elkowar** [comment]
ahh yea no, no variables in the window tag. What you can do is use the command line arguments for the eww open command, where you can specify window geometry and stuff for the window specifically the moment you open it. check the output of `eww open --help`

**Comment by ckp95** [comment]
Ah okay, yes that works.
```
eww open --pos=450x200 --size=100x100 simple_box
```
Btw you might want to add to the `--help` output how the position and size arguments are meant to be formatted; I had to figure it out from the error messages.

**Comment by godalming123** [comment]
You could play around wtth making a giant window and then styling it to have a transparent background and an inner element with varying size with the background

**Comment by ModProg** [comment]
Does this still work? And does it work on wayland?

**Comment by godalming123** [comment]
> Does this still work? And does it work on wayland?

It should still work although I have not checked however the cursor events will be sent to the eww window which means that when it is open you would not be able to use other windows

**Comment by Wallboy** [comment]
> Ah okay, yes that works.
> 
> ```
> eww open --pos=450x200 --size=100x100 simple_box
> ```
> 
> Btw you might want to add to the `--help` output how the position and size arguments are meant to be formatted; I had to figure it out from the error messages.

I can't get this to work at all. The window always opens up at the top left of my monitor no matter what values I input.
